src/caffe/util/gpu_memory.cpp

```diff
@@ -37,7 +37,7 @@ void GPUMemory::Manager::init(const vector<int>& gpus, Mode m, bool debug) {
       // Just in case someone installed 'no cleanup' arena before
       delete cub_allocator_;
       cub_allocator_ = new cub::CachingDeviceAllocator(BIN_GROWTH, MIN_BIN,
-          MAX_BIN, MAX_CACHED_BYTES, false, debug_);
+          MAX_BIN, MAX_CACHED_BYTES, true, debug_);
```

Comment:
That was the problem. We do clean what we can. The rest is cleaned by driver. ~Scope doesn't help either. All this is caoused by different life times for objects working with different GPUs. Look here: https://github.com/NVIDIA/caffe/blob/caffe-0.15/src/caffe/parallel.cpp#L316-L319 This loop runs from 1, not from 0. We treat GPU 0 differently. I don't claim it's wrong. But it makes memory management more interesting. :)